chore(docs): Add tutorial (#62)
Soluto/dynamico: tutorial.md

# Dynamico Tutorial

In this tutorial you're going to use a demo registry we've set up to host and serve components. The goal here is to try out the work flow with dynamico, and exprience how easy it is to integrate and use dynamico.

> We are going to create a new react app and integrate dynamico client into it. After that we're going to create a new component, developing it locally and publish it to a server and see it work in production mode. When we're finished with that we will update the component and see the changes applied in your app.

Reviewer: ```suggestion
We are going to create a new react app or use your own and integrate dynamico client into it. After that we're going to create a new component, developing it locally and publish it to a server and see it work in production mode. Last step will be updating the component, bump the version and see the changes applied in your app.
```

This guide assumes you have a node version compatible with `create-react-app` (node 10 works fine), yarn and npx available.

## Host App

> Let's start with creating an app:

Reviewer: ```suggestion
Let's start with creating an app (skip if you're using your own app)
```

```bash

> $ npx create-react-app dynamico-demo

Reviewer: ```suggestion
$ npx create-react-app dynamico-demo
$ cd dynamico-demo
```

```

Now install the dependencies:
```bash

> $ cd dynamico-demo

Reviewer: ```suggestion
```

& yarn add @dynamico/core @dynamico/react
```

Open `App.js` and add these import statements to the top of the file:
```javascript
import { DynamicoClient } from '@dynamico/core';
import { DynamicoProvider } from '@dynamico/react';
```

> Here we import both dynamico client which is the client side 

Reviewer: ```suggestion
We import both dynamico client which is the client side 
```

library responsible for extracting components from the code 

> and dynamico provider which is used to pass the client along on the context to so you can use dynamic components just like a regular react component.

Reviewer: ```suggestion
and dynamico provider which is used to pass the client along on the context so you can use dynamic components just like a regular react component.
```

> We're also going to need ReactDOM available in this scope so let's go ahead and import that too:

Reviewer: why?

Author: it's one of the default dependencies we add in plop when creating a component... so for this to Just Work ™️  you have to add react-dom. We can add moment or lodash for the update step in the tutorial.

Reviewer: We should definitely remove it from there :X

```javascript
import * as ReactDOM from 'react-dom';
```

> Now we're going to initialize the client. It requires a few mandatory parameters. The first is the URL of the registry to be used in production. You can use the demo registry we've set up. the url is: `https://dynamico-demo-registry.soluto.io`.

Reviewer: ```suggestion
Now we're going to initialize the client. It requires a few mandatory parameters. The first is the URL of the registry to be used in production. You can use the demo registry we've set up. the url is: `https://dynamico-demo-registry.soluto.io`.

To read more about our registry go to our [docs]()
```

> Note: Components you publish to this registry are only available for a day, and are deleted a day after they're published.

> The next is the storage to use for local caching of components. The API for the cache matches localStorage's API so you can just pass it as the cache.

Reviewer: ```suggestion
The next is the storage to use for local caching of components. The API for the cache matches `localStorage`'s API so you can just pass it as the cache.
```

> Last is the dependencies we provide from the host app to components. We'll pass in just React and ReactDOM for now.

Reviewer: ```suggestion
Last is the dependencies we provide from the host app to the components. We'll pass in just React and ReactDOM for now.
```

For the resolvers we'll define this object:
```javascript
const resolvers = {
  react: React,
  'react-dom': ReactDOM
};
```

And now we have everything we need to initialize the client. Add these lines to your code:
```javascript
const dynamicoClient = new DynamicoClient({
  url: 'https://dynamico-demo-registry.soluto.io',
  dependencies: {
    versions: dependencies,
    resolvers
  },
  cache: localStorage
});
```

Now we're going to initialize our react provider. Doing it is as easy as adding a single component to wrap the App's top level component:
```javascript
function App() {
  return (
      <DynamicoProvider client={dynamicoClient}>
        <div className="App">
        <header className="App-header">
            <img src={logo} className="App-logo" alt="logo" />
            <p>
            Edit <code>src/App.js</code>...
            </p>
            <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
            >
            Learn React
            </a>
        </header>
        </div>
    </DynamicoProvider>
  );
}
```

Now we can try and run the app:
```bash
$ yarn start
```

> And you'll see CRA's default view.

Reviewer: ```suggestion
And nothing actually change, let's add dynamic components!
```

> Let's add a dynamic component. Open App.js again and add dynamico to the import statement from @dynamico/react:

Reviewer: I'd say, let's create a new file

Author: I want to keep it simple and a minimal number of steps. Creating a new file requires more work and things to get wrong... I know adding a file and importing it is simple but this is a tutorial, not best practices guide. Am I making sense?

```javascript
import { DynamicoProvider, dynamico } from '@dynamico/react';
```

Now add the code that actually creates the dynamic component:
```javascript
const DynamicComponent = dynamico('dynamiccomponent', {
  fallback: <div>Loading...</div>
});
```

Here we tell dynamico to get a component called `dynamiccomponent` and show `<div>Loading...</div>` In the meanwhile or if anything fails. There are more ways to retrieve dynamic components but this is the simplest way if you're using React.
> Note: The client doesn't try to download code until you actually try to render the component.

Let's try to render this component. Change the App's component code to look like this:
```javascript
function App() {
  return (
    <DynamicoProvider client={dynamicoClient}>
      <div className="App">

>       <DynamicComponent />

Reviewer: ```suggestion
        <DynamicComponent />
```

        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <p>
            Edit <code>src/App.js</code> and save to reload.
          </p>
          <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
        </header>
      </div>
    </DynamicoProvider>
  );
}
```

We added our `DynamicComponent` right above the header tag. If we run the app again we see that some bar is added at the top of the screen that says `Loading...`. This is our fallback component. This happens because the client tries to get our `dynamiccomponent` from the registry but we didn't publish it yet.

## Creating a component

> To create a component you're going to need our cli:

Reviewer: link to docs
```suggestion
To create a component you're going to need our [cli]():
```

  <div>
    <span>This is fun!</span>
  </div>
);
```

Note that as you save your app gets updated automatically.

Now let's publish the component. Got to your terminal, stop the dcm development server, and open `dcmconfig.js`. You'll see that this file exports an object with a `registry` property. It expects to have a URL of an available registry. You should use the same URL you used when you set up dynamico client in the app:
```javascript
exports.default = {
  registry: 'https://dynamico-demo-registry.soluto.io'
};
```

Now go back to your terminal in your component's folder and run:
```bash
$ dcm publish
```

You should see:
```bash
17:29:16 [dcm] info: Publishing...
success Bundled ~/dynamiccomponent/index.jsx in cjs format (559ms)
┌─────────────────────────────┐
│File           Size   Gzipped│
│dist/index.js  290 B  203 B  │
└─────────────────────────────┘
17:29:20 [dcm] info: Successfully published Dynamiccomponent@1.0.0
```

Now let's try and use it in the app. Open App.js and remove devMode property from the dynamic component's creation:
```javascript
const DynamicComponent = dynamico('dynamiccomponent', {
  fallback: <div>Loading...</div>
});
```

> Run the app again and you can see that it takes a bit longer for the component to load but eventually you'll see your component's code!

Reviewer: also maybe we should add that next time it will be fetched from localStorage
